### ISSUE TITLE
psm3_src_chksum.h issue may break build

### ISSUE THREAD

**Issue by chuckcranor** [ofiwg/libfabric#6773]
There is a dependency issue with creating psm3_src_chksum.h that shows up if you compile libfabric without the parallel job "-j" flag to make.  i.e. if you do this:

<pre>
git clone https://github.com/ofiwg/libfabric.git
cd libfabric 
./autogen.sh 
./configure
#  XXX NOTE: to trigger error, do not use "-j" with make
make
</pre>

Then you get this error:
<pre>
  CC       prov/psm3/src/libpsm3_la-psmx3_tagged.lo
  CC       prov/psm3/src/libpsm3_la-psmx3_trx_ctxt.lo
  CC       prov/psm3/src/libpsm3_la-psmx3_util.lo
  CC       prov/psm3/src/libpsm3_la-psmx3_wait.lo
  CC       prov/psm3/src/libpsm3_la-psm3_revision.lo
prov/psm3/src/psm3_revision.c:1:10: fatal error: psm3_src_chksum.h: No such file or directory
    1 | #include "psm3_src_chksum.h"
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [Makefile:13630: prov/psm3/src/libpsm3_la-psm3_revision.lo] Error 1
make[1]: Leaving directory '/tmp/ofi/libfabric'
make: *** [Makefile:6172: all] Error 2
% find . -name psm3_src\*
% 
</pre>

If you rerun a plain "make" it will still fail with the same error.   If you throw in a "make -j 2" it will fail, but then the next plain make will work.   e.g.

<pre>
% make
make  all-am
make[1]: Entering directory '/tmp/ofi/libfabric'
  CC       prov/psm3/src/libpsm3_la-psm3_revision.lo
prov/psm3/src/psm3_revision.c:1:10: fatal error: psm3_src_chksum.h: No such file or directory
    1 | #include "psm3_src_chksum.h"
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [Makefile:13630: prov/psm3/src/libpsm3_la-psm3_revision.lo] Error 1
make[1]: Leaving directory '/tmp/ofi/libfabric'
make: *** [Makefile:6172: all] Error 2
% 
% make
make  all-am
make[1]: Entering directory '/tmp/ofi/libfabric'
  CC       prov/psm3/src/libpsm3_la-psm3_revision.lo
prov/psm3/src/psm3_revision.c:1:10: fatal error: psm3_src_chksum.h: No such file or directory
    1 | #include "psm3_src_chksum.h"
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [Makefile:13630: prov/psm3/src/libpsm3_la-psm3_revision.lo] Error 1
make[1]: Leaving directory '/tmp/ofi/libfabric'
make: *** [Makefile:6172: all] Error 2
% 
% make -j 2
make  all-am
make[1]: Entering directory '/tmp/ofi/libfabric'
  CC       prov/psm3/src/libpsm3_la-psm3_revision.lo
prov/psm3/src/psm3_revision.c:1:10: fatal error: psm3_src_chksum.h: No such file or directory
    1 | #include "psm3_src_chksum.h"
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [Makefile:13630: prov/psm3/src/libpsm3_la-psm3_revision.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/tmp/ofi/libfabric'
make: *** [Makefile:6172: all] Error 2
% 
% make
make  all-am
make[1]: Entering directory '/tmp/ofi/libfabric'
  CC       prov/psm3/src/libpsm3_la-psm3_revision.lo
  CC       prov/psm3/psm3/libpsm3i_la-psm.lo
  CC       prov/psm3/psm3/libpsm3i_la-psm_am.lo
  CC       prov/psm3/psm3/libpsm3i_la-psm_context.lo
  CC       prov/psm3/psm3/libpsm3i_la-psm_diags.lo
  CC       prov/psm3/psm3/libpsm3i_la-psm_ep.lo
</pre>

It seems like the rule to build psm3_src_chksum.h is not firing early enough, but if you run "make -j" the parallel jobs inadvertently work around the underlying problem in the build system... but I think the underlying problem should be fixed.

**Comment by shefty** [comment]
@acgoldma - please look into this, thanks.

**Comment by acgoldma** [comment]
Looks like the checksum was not properly being handle in the dependencies. AutoMake does not really have a clean way of doing this other than BUILT_SOURCES

This seems to work:
```
diff --git a/prov/psm3/Makefile.include b/prov/psm3/Makefile.include
index 11bb164d9..474550d19 100644
--- a/prov/psm3/Makefile.include
+++ b/prov/psm3/Makefile.include
@@ -282,9 +282,10 @@ libpsm3_la_DEPENDENCIES = \
        prov/psm3/src/psm3_src_chksum.h \
        libpsm3i.la

+BUILT_SOURCES = prov/psm3/src/psm3_src_chksum.h
 CLEANFILES = prov/psm3/src/psm3_src_chksum.h
 prov/psm3/src/psm3_src_chksum.h: Makefile $(chksum_srcs)
-       @ chksum=`for file in $(chksum_srcs); do cat $(top_srcdir)/$$file; done | sha1sum | cut -d' ' -f 1`; \
+       $(AM_V_GEN) chksum=`for file in $(chksum_srcs); do cat $(top_srcdir)/$$file; done | sha1sum | cut -d' ' -f 1`; \
        if ! grep -q $$chksum prov/psm3/src/psm3_src_chksum.h 2>/dev/null; then \
                echo "#define PSMX3_SRC_CHECKSUM \"$$chksum\"" > prov/psm3/src/psm3_src_chksum.h; \
                echo "#define PSMX3_BUILD_TIMESTAMP \"`date`\"" >> prov/psm3/src/psm3_src_chksum.h; \
```

Sample Output of `make`:
```
# make
  GEN      prov/psm3/src/psm3_src_chksum.h
make  all-am
make[1]: Entering directory '/root/acg/libfabric'
  CC       util/info.o
  CC       src/libfabric_la-fabric.lo
...
```

Only the `BUILT_SOURCES=` line is required, the other is just to add the output to `make`

**Comment by chuckcranor** [comment]
I just tried the BUILT_SOURCES patch and it worked for me too, with the GEN being the first thing done by make (same as your sample output).

**Comment by shefty** [comment]
Fix from here isolated and merged into v1.13.0rc1
